conda/common/logic.py

```diff
@@ -128,92 +309,132 @@ def Prevent(self, what, *args):
     def Require(self, what, *args):
         return what.__get__(self, Clauses)(*args, polarity=True, name=False)
 
-    def Not_(self, x, polarity=None):
+    def Not_(self, x, polarity=None, add_new_clauses=False):
         return (not x) if type(x) is bool else -x
 
     def Not(self, x, polarity=None, name=None):
         return self.Eval_(self.Not_, (x,), polarity, name)
 
-    def And_(self, f, g, polarity=None):
+    def And_(self, f, g, polarity, add_new_clauses=False):
         if f is False or g is False:
             return False
         if f is True:
             return g
-        if g is True or f is g:
+        if g is True:
+            return f
+        if f == g:
             return f
-        if f is -g:
+        if f == -g:
             return False
         if g < f:
             f, g = g, f
+        if add_new_clauses:
```

Comment:
This code duplication pains me, but is needed if we run CPython :unamused:
https://github.com/conda/conda/pull/7676/files#r211312654